### PR TITLE
Speechd bridge (nodejs)

### PR DESCRIPTION
**Pull request by andrewn**
An implementation of #39 using JavaScript. It autospawns a `speech-dispatcher` instance in a similar way to the Python API client and communicates over a UNIX socket. I was concerned that [the SSIP protocol](https://devel.freebsoft.org/doc/speechd/ssip.html) be a bit difficult to work with but it seems good. 

I do prefer this approach as nodejs is better suited to this kind of async stuff. We can also reuse the upcoming WebSocket client and any other infrastructure we create for managing node services.

### CODE REVIEW

File: services/speech/src/speechd/connect.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const spawnServer = require("./spawnServer");
+const { ConnectionError } = require("./errors");
+
+const connectToSocketOrNull = async params => {
+  try {
+    return await connectToSocket(params);
+  } catch (e) {
+    const isConnectionError = ["ENOENT", "ECONNREFUSED"].includes(e.code);
+    if (isConnectionError) {
+      return null;
```
Comment (pixelblend) [inline services/speech/src/speechd/connect.js]: I'm not sure why this returns `null` for a connection error, when it's then conditionally raised later on. Why not have this function throw the connectionError, which can then be optionally caught and suppressed by the function that calls it?

I think that would make this function clearer, and the callee doesn't have to throw it's own exceptions.

File: services/speech/src/speechd/environment.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+const exec = require("child_process").exec;
+
+const getExecutablePath = async () =>
+  new Promise((resolve, reject) => {
+    exec(`which speech-dispatcher`, (error, stdout) => {
```
Comment (pixelblend) [inline services/speech/src/speechd/environment.js]: I'd never expect this to throw an error, unless `which` wasn't found in the `PATH`. I think it's more likely to return with an empty `stdout` if the `speech-dispatcher` isn't found.

File: services/speech/src/speechd/errors.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class ConnectionError extends Error {
+  constructor() {
+    super("Cannot connect to speech-dispatcher server");
```
Comment (pixelblend) [inline services/speech/src/speechd/errors.js]: Custom Errors 😍

File: services/speech/src/speechd/environment.js (hunk 2 above)
Comment (pixelblend) [reply services/speech/src/speechd/environment.js]: ... I am *dead wrong* about this. `which` will return an exit code of `1` if the binary isn't found in `$PATH`.

**Comment by andrewn** [comment]
@pixelblend, this is ready for re-review please.

I've addressed your previous comments and added some more functionality:

> I'm not sure why this returns null for a connection error, when it's then conditionally raised later on. Why not have this function throw the connectionError, which can then be optionally caught and suppressed by the function that calls it?

Fixed in 1de6762.

> `which` will return an exit code of 1 if the binary isn't found in $PATH.

Fixed in 339c3a.

I've also added the ability listing supported voice types (8fc24982) and specifying the voice type when sending the `speak` command (613750e4).

File: services/speech/src/handleMessage.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module.exports = (speech, broker) => async (topic, payload) => {
+  try {
+    switch (topic) {
+      case "speech.command.speak":
```
Comment (pixelblend) [inline services/speech/src/handleMessage.js]: Don't forget to alter these to be `/` in the code and docs :)

**Pull request merged**